src/main/groovy/wooga/gradle/unity/UnityTask.groovy

```diff
@@ -0,0 +1,188 @@
+import wooga.gradle.unity.utils.ForkTextStream
+import wooga.gradle.unity.utils.UnityVersionManager
+
+abstract class UnityTask extends DefaultTask
+        implements UnitySpec,
+                UnityCommandLineSpec,
+                ArgumentsSpec {
+
+    UnityTask() {
+        setCommandLineOptionDefaults()
+
+        // When this task is executed, we query the arguments to pass
+        // onto the Unity process here. We generate a sequence of Unity's command line options
+        // and also an additional one for our custom use
+        wooga_gradle_unity_traits_ArgumentsSpec__arguments = project.provider({
+            List<String> arguments = new ArrayList<String>()
+
+            // Add command line options
+            def unityCmdLineOptions = getUnityCommandLineOptions()
+            unityCommandLineOptions.each({ arguments.add(it) })
+
+            // Add additional arguments
+            if (additionalArguments.present) {
+                additionalArguments.get().each {
+                    arguments << it
+                }
+            }
+
+            arguments
+        })
+
+        wooga_gradle_unity_traits_ArgumentsSpec__additionalArguments = project.objects.listProperty(String)
+
+    }
+
+    @TaskAction
+    void exec() {
+        ExecResult execResult = project.exec(new Action<ExecSpec>() {
+            @Override
+            void execute(ExecSpec exec) {
+
+                if (!unityPath.present) {
+                    throw new GradleException("Unity path is not set")
+                }
+
+                preExecute()
+
+                def unityPath = unityPath.get().asFile.absolutePath
+                def unityArgs = arguments.get()
+
+                OutputStream outputStream = getOutputStream()
+
+                exec.with {
+                    executable unityPath
+                    args = unityArgs
+                    standardOutput = outputStream
+                    ignoreExitValue = true
+                }
+            }
+        })
+
+        execResult.assertNormalExitValue()
+        postExecute(execResult)
+    }
+
+    /**
+     * Invoked before the task executes the Unity process
+     */
+    protected void preExecute() {
+        if (projectDirectory.present && !isSet(UnityCommandLineOption.projectPath)) {
+            projectPath = projectDirectory.get().asFile.path
+        }
+    }
+
+    /**
+     * Invoked after the task executes the Unity process
+     * @param result The result of the Unity execution
+     */
+    protected void postExecute(ExecResult result) {
+    }
+
+    /**
+     * Sets the default command line options for the task
+     */
+    protected void setCommandLineOptionDefaults() {
+        batchMode.convention(true)
+        getCommandLineOption(UnityCommandLineOption.quit).enabled.convention(true)
+        getCommandLineOption(UnityCommandLineOption.logFile).enabled.convention(true)
+//        quit = true
+//        logFile = true
+    }
```

Comment:
I'm not sure what the call to convention is doing here. This is the base UnityTask. These default values should not be coded in the Task itself but rather in the Plugin as default values for all tasks of type UnityTask except for subclasses that need to override this.

Reply:
Good point; it had been done for the other variables.